--- a/lib/autoconf/fortran.m4
+++ b/lib/autoconf/fortran.m4
@@ -665,7 +665,7 @@ while test $[@%:@] != 1; do
 	-lang* | -lcrt*.o | -lc | -lgcc* | -lSystem | -libmil | -little \
 	  |-LANG:=* | -LIST:* | -LNO:* | -link)
 	  ;;
-	-lkernel32 | -lmingw* | -lmoldname)
+	-lkernel32)
 	  # Ignore this library only on Windows-like systems.
 	  case $host_os in
 	  cygwin* | msys* | mingw* | windows*) ;;
-- 


--- a/bin/autoreconf.in
+++ b/bin/autoreconf.in
@@ -440,6 +440,8 @@ sub autoreconf_current_directory ($)
   # ---------------------- #
 
   my $uses_autoconf;
+  my $uses_liblink;
+  my $uses_iconv;
   my $uses_gettext;
   if (-f $configure_ac)
     {
@@ -449,8 +451,13 @@ sub autoreconf_current_directory ($)
 	  s/#.*//;
 	  s/dnl.*//;
 	  $uses_autoconf = 1 if /AC_INIT/;
+	  $uses_liblink = 1 if /AC_LIB_HAVE_LINKFLAGS/;
+	  $uses_liblink = 1 if /AC_LIB_LINKFLAGS/;
+	  $uses_liblink = 1 if /AC_LIB_LINKFLAGS_FROM_LIBS/;
+	  $uses_iconv = 1 if /AM_ICONV/;
 	  # See below for why we look for gettext here.
-	  $uses_gettext = 1  if /^AM_GNU_GETTEXT_(?:REQUIRE_)?VERSION/;
+	  $uses_gettext = 1 if /AM_GNU_GETTEXT/;
+	  $uses_gettext = 1 if /AM_PO_SUBDIRS/;
 	}
       if (!$uses_autoconf)
         {
@@ -485,7 +492,7 @@ sub autoreconf_current_directory ($)
   # Actually, it is even more restrictive, as it greps for
   # '^AM_GNU_GETTEXT_(REQUIRE_)?VERSION('.  We did this above, while
   # scanning configure.ac.
-  if (!$uses_gettext)
+  if (!$uses_liblink && !$uses_iconv && !$uses_gettext)
     {
       verb "$configure_ac: not using Gettext";
     }
@@ -495,8 +502,21 @@ sub autoreconf_current_directory ($)
     }
   else
     {
-      xsystem_hint ("autopoint is needed because this package uses Gettext",
-                    $autopoint);
+      if ($uses_gettext)
+        {
+          xsystem_hint ("autopoint is needed because this package uses Gettext",
+                        $autopoint);
+        }
+      elsif ($uses_iconv)
+        {
+          xsystem_hint ("autopoint is needed because this package uses AM_ICONV",
+                        $autopoint);
+        }
+      elsif ($uses_liblink)
+        {
+          xsystem_hint ("autopoint is needed because this package uses AC_LIB_LINKFLAGS",
+                        $autopoint);
+        }
     }
 
 
@@ -576,6 +596,8 @@ sub autoreconf_current_directory ($)
   # from the final autoconf invocation.
   my $aux_dir;
   my @aux_files;
+  my $uses_liblink_via_traces;
+  my $uses_iconv_via_traces;
   my $uses_gettext_via_traces;
   my $uses_libtool;
   my $uses_intltool;
@@ -603,6 +625,11 @@ sub autoreconf_current_directory ($)
                'AM_PROG_LIBTOOL',
                'LT_INIT',
                'LT_CONFIG_LTDL_DIR',
+               'AC_LIB_LINKFLAGS',
+               'AC_LIB_HAVE_LINKFLAGS',
+               'AC_LIB_LINKFLAGS_FROM_LIBS',
+               'GUILE_FLAGS',
+               'AM_ICONV',
                'AM_GNU_GETTEXT',
                'AM_INIT_AUTOMAKE',
                'GTK_DOC_CHECK',
@@ -617,6 +644,15 @@ sub autoreconf_current_directory ($)
       $aux_dir = $args[0]           if $macro eq "AC_CONFIG_AUX_DIR";
       push @aux_files, $args[0]     if $macro eq "AC_REQUIRE_AUX_FILE";
       $uses_autoconf = 1            if $macro eq "AC_INIT";
+      # Here we need to explicitly test for GUILE_FLAGS, because in the
+      # typical situation where configure.ac invokes GUILE_FLAGS,
+      # guile.m4 defines GUILE_FLAGS and uses AC_LIB_LINKFLAGS_FROM_LIBS,
+      # the traces still for AC_LIB_LINKFLAGS_FROM_LIBS are empty.
+      $uses_liblink_via_traces = 1  if $macro eq "AC_LIB_LINKFLAGS"
+                                       || $macro eq "AC_LIB_HAVE_LINKFLAGS"
+                                       || $macro eq "AC_LIB_LINKFLAGS_FROM_LIBS"
+                                       || $macro eq "GUILE_FLAGS";
+      $uses_iconv_via_traces = 1    if $macro eq "AM_ICONV";
       $uses_gettext_via_traces = 1  if $macro eq "AM_GNU_GETTEXT";
       $uses_libtool = 1             if $macro eq "AC_PROG_LIBTOOL"
                                        || $macro eq "AM_PROG_LIBTOOL"
@@ -741,6 +777,42 @@ sub autoreconf_current_directory ($)
     }
 
 
+  # --------------------------------------------- #
+  # Running autopoint, if not already run above.  #
+  # --------------------------------------------- #
+
+  if (!$uses_liblink_via_traces && !$uses_iconv_via_traces && !$uses_gettext_via_traces)
+    {
+      verb "$configure_ac: not needing autopoint";
+    }
+  elsif (!$install)
+    {
+      verb "$configure_ac: not running autopoint: --install not given";
+    }
+  elsif ($uses_liblink || $uses_iconv || $uses_gettext)
+    {
+      verb "$configure_ac: not running autopoint a second time";
+    }
+  else
+    {
+      if ($uses_gettext_via_traces)
+        {
+          xsystem_hint ("autopoint is needed because this package uses Gettext",
+                        $autopoint);
+        }
+      elsif ($uses_iconv_via_traces)
+        {
+          xsystem_hint ("autopoint is needed because this package uses AM_ICONV",
+                        $autopoint);
+        }
+      elsif ($uses_liblink_via_traces)
+        {
+          xsystem_hint ("autopoint is needed because this package uses AC_LIB_LINKFLAGS",
+                        $autopoint);
+        }
+    }
+
+
   # ------------------- #
   # Rerunning aclocal.  #
   # ------------------- #
-- 


--- a/bin/autoreconf.in
+++ b/bin/autoreconf.in
@@ -489,6 +489,7 @@ sub autoreconf_current_directory ($)
   # Actually, it is even more restrictive, as it greps for
   # '^AM_GNU_GETTEXT_(REQUIRE_)?VERSION('.  We did this above, while
   # scanning configure.ac.
+  my $invoked_autopoint;
   if (!$uses_liblink && !$uses_iconv && !$uses_gettext)
     {
       verb "$configure_ac: not using Gettext";
@@ -514,6 +515,7 @@ sub autoreconf_current_directory ($)
           xsystem_hint ("autopoint is needed because this package uses AC_LIB_LINKFLAGS",
                         $autopoint);
         }
+      $invoked_autopoint = 1;
     }
 
 
@@ -593,6 +595,7 @@ sub autoreconf_current_directory ($)
   # from the final autoconf invocation.
   my $aux_dir;
   my @aux_files;
+  my @macro_dirs;
   my $uses_liblink_via_traces;
   my $uses_iconv_via_traces;
   my $uses_gettext_via_traces;
@@ -616,6 +619,8 @@ sub autoreconf_current_directory ($)
                'AC_CONFIG_AUX_DIR',
                'AC_CONFIG_HEADERS',
                'AC_CONFIG_SUBDIRS',
+               'AC_CONFIG_MACRO_DIR_TRACE',
+               'AC_CONFIG_MACRO_DIR',
                'AC_INIT',
                'AC_REQUIRE_AUX_FILE',
                'AC_PROG_LIBTOOL',
@@ -640,6 +645,8 @@ sub autoreconf_current_directory ($)
       my ($macro, @args) = split (/::/);
       $aux_dir = $args[0]           if $macro eq "AC_CONFIG_AUX_DIR";
       push @aux_files, $args[0]     if $macro eq "AC_REQUIRE_AUX_FILE";
+      push @macro_dirs, $args[0]    if $macro eq "AC_CONFIG_MACRO_DIR_TRACE"
+                                       || $macro eq "AC_CONFIG_MACRO_DIR";
       $uses_autoconf = 1            if $macro eq "AC_INIT";
       # Here we need to explicitly test for GUILE_FLAGS, because in the
       # typical situation where configure.ac invokes GUILE_FLAGS,
@@ -679,6 +686,15 @@ sub autoreconf_current_directory ($)
 	}
     }
 
+  # When configure.ac does not specify a macro dir, autopoint has
+  # created a directory named 'm4' and copied its *.m4 files into it.
+  # In this case, we need to rerun 'aclocal', in order to pass it the
+  # options "-I m4" this time.
+  if ($invoked_autopoint && !@macro_dirs)
+    {
+      $rerun_aclocal = 1;
+    }
+
   # Gettext consistency checks.
   # Some projects intentionally don't call AM_GNU_GETTEXT_(REQUIRE_)VERSION
   # because they have all of the gettext infrastructure checked into version
@@ -807,6 +823,8 @@ sub autoreconf_current_directory ($)
           xsystem_hint ("autopoint is needed because this package uses AC_LIB_LINKFLAGS",
                         $autopoint);
         }
+      $invoked_autopoint = 1;
+      $rerun_aclocal = 1;
     }
 
 
@@ -824,6 +842,12 @@ sub autoreconf_current_directory ($)
 	}
       else
 	{
+	  # When configure.ac does not specify a macro dir, autopoint has
+	  # created a directory named 'm4' and copied its *.m4 files into it.
+	  if ($invoked_autopoint && !@macro_dirs)
+	    {
+	      $aclocal_flags .= " -I m4";
+	    }
 	  xsystem ("$aclocal $aclocal_flags");
 	}
     }
-- 
