--- input.c
+++ input.c
@@ -261,13 +261,16 @@ rl_gather_tyi (void)
   input = 0;
   tty = fileno (rl_instream);
 
-  /* Move this up here to give it first shot, but it can't set chars_avail */
+  /* Move this up here to give it first shot, but it can't set chars_avail,
+     so we assume a single character is available. */
   /* XXX - need rl_chars_available_hook? */
   if (rl_input_available_hook)
     {
       result = (*rl_input_available_hook) ();
       if (result == 0)
         result = -1;
+      else
+        chars_avail = 1;
     }
 
 #if defined (HAVE_PSELECT) || defined (HAVE_SELECT)
@@ -285,6 +288,7 @@ rl_gather_tyi (void)
 #endif
       if (result <= 0)
 	return 0;	/* Nothing to read. */
+      result = -1;	/* there is something, so check how many chars below */
     }
 #endif
 
--- patchlevel
+++ patchlevel
@@ -1,3 +1,3 @@
 # Do not edit -- exists only for use by patch
 
-0
+1


--- readline.h
+++ readline.h
@@ -409,7 +409,7 @@ extern void rl_activate_mark (void);
 extern void rl_deactivate_mark (void);
 extern int rl_mark_active_p (void);
 
-extern int rl_message (const char *, ...)  __attribute__((__format__ (printf, 1, 2)));
+extern int rl_message (const char *, ...)  __rl_attribute__((__format__ (printf, 1, 2)));
 
 extern int rl_show_char (int);
 
--- rlstdc.h
+++ rlstdc.h
@@ -36,10 +36,10 @@
 #  endif
 #endif
 
-#ifndef __attribute__
-#  if __GNUC__ < 2 || (__GNUC__ == 2 && __GNUC_MINOR__ < 8)
-#    define __attribute__(x)
-#  endif
+#if defined(__GNUC__) && __GNUC__ >= 2
+#  define __rl_attribute__(x) __attribute__(x)
+#else
+#  define __rl_attribute__(x)
 #endif
 
 #endif /* !_RL_STDC_H_ */


--- history.h
+++ history.h
@@ -32,6 +32,7 @@ extern "C" {
 #  include "rlstdc.h"
 #  include "rltypedefs.h"
 #else
+#  include <stdio.h>
 #  include <readline/rlstdc.h>
 #  include <readline/rltypedefs.h>
 #endif
--- readline.h
+++ readline.h
@@ -32,6 +32,7 @@ extern "C" {
 #  include "keymaps.h"
 #  include "tilde.h"
 #else
+#  include <stdio.h>
 #  include <readline/rlstdc.h>
 #  include <readline/rltypedefs.h>
 #  include <readline/keymaps.h>


--- configure
+++ configure
@@ -3808,7 +3808,7 @@ ac_clean_files="$ac_clean_files conftest
 # the compiler is broken, or we cross compile.
 { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking whether we are cross compiling" >&5
 printf %s "checking whether we are cross compiling... " >&6; }
-if test "$cross_compiling" != yes; then
+if false && test "$cross_compiling" != yes; then
   { { ac_try="$ac_link"
 case "(($ac_try" in
   *\"* | *\`* | *\\*) ac_try_echo=\$ac_try;;


--- support/shobj-conf
+++ support/shobj-conf
@@ -128,7 +128,7 @@ linux*-*|gnu*-*|k*bsd*-gnu-*|freebsd*-ge
 	SHOBJ_LD='${CC}'
 	SHOBJ_LDFLAGS='-shared -Wl,-soname,$@'
 
-	SHLIB_XLDFLAGS='-Wl,-rpath,$(libdir) -Wl,-soname,`basename $@ $(SHLIB_MINOR)`'
+	SHLIB_XLDFLAGS='-Wl,-soname,`basename $@ $(SHLIB_MINOR)`'
 	SHLIB_LIBVERSION='$(SHLIB_LIBSUFF).$(SHLIB_MAJOR)$(SHLIB_MINOR)'
 	;;
 


--- examples/rlfe/rlfe.c
+++ examples/rlfe/rlfe.c
@@ -152,21 +152,27 @@ struct termios orig_term;
 static pid_t child = -1;
 
 static void
-sig_child (int signo)
+finish_up()
 {
-  int status;
-  wait (&status);
   if (hist_file != 0)
     {
       write_history (hist_file);
       if (hist_size)
 	history_truncate_file (hist_file, hist_size);
     }
-  DPRINT0 ("(Child process died.)\n");
   tcsetattr(STDIN_FILENO, TCSANOW, &orig_term);
   exit (0);
 }
 
+static void
+sig_child (int signo)
+{
+  int status;
+  wait (&status);
+  DPRINT0 ("(Child process died.)\n");
+  finish_up();
+}
+
 volatile int propagate_sigwinch = 0;
 
 /* sigwinch_handler
@@ -708,8 +714,7 @@ main(int argc, char** argv)
 	  if (count <= 0)
 	    {
 	      DPRINT0 ("(Connection closed by foreign host.)\n");
-	      tcsetattr(STDIN_FILENO, TCSANOW, &orig_term);
-	      exit (0);
+              finish_up();
 	    }
 	  old_count = buf_count;
 


--- examples/rlfe/pty.c
+++ examples/rlfe/pty.c
@@ -131,7 +131,7 @@ InitPTY(int f)
 {
   if (f < 0)
     return;
-#if defined(I_PUSH) && defined(HAVE_SVR4_PTYS) && !defined(sgi) && !defined(linux) && !defined(__osf__) && !defined(M_UNIX)
+#if defined(I_PUSH) && defined(HAVE_SVR4_PTYS) && !defined(sgi) && !defined(linux) && !defined(__osf__) && !defined(M_UNIX) && !defined(__FreeBSD_kernel__)
   if (ioctl(f, I_PUSH, "ptem"))
     Panic(errno, "InitPTY: cannot I_PUSH ptem");
   if (ioctl(f, I_PUSH, "ldterm"))


--- examples/rlfe/configure.ac
+++ examples/rlfe/configure.ac
@@ -273,8 +273,8 @@ __sorry_hpux_libcurses_is_totally_broken
 tgetent();
 #endif
 ],,
-LIBS="-ltermcap $olibs"
-AC_MSG_CHECKING(libtermcap)
+LIBS="-ltinfo $olibs"
+AC_MSG_CHECKING(libtinfo)
 AC_TRY_LINK([extern char tgetent(void);],tgetent();,,
 LIBS="-ltermlib $olibs"
 AC_MSG_CHECKING(libtermlib)
--- examples/rlfe/configure
+++ examples/rlfe/configure
@@ -4448,9 +4448,9 @@ if ac_fn_c_try_link "$LINENO"
 then :
 
 else case e in #(
-  e) LIBS="-ltermcap $olibs"
-{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking libtermcap" >&5
-printf %s "checking libtermcap... " >&6; }
+  e) LIBS="-ltinfo $olibs"
+{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking libtinfo" >&5
+printf %s "checking libtinfo... " >&6; }
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 extern char tgetent(void);


--- aclocal.m4
+++ aclocal.m4
@@ -932,8 +932,8 @@ _bash_needmsg=
 fi
 AC_CACHE_VAL(bash_cv_termcap_lib,
 [AC_CHECK_FUNC(tgetent, bash_cv_termcap_lib=libc,
-  [AC_CHECK_LIB(termcap, tgetent, bash_cv_termcap_lib=libtermcap,
-    [AC_CHECK_LIB(tinfo, tgetent, bash_cv_termcap_lib=libtinfo,
+  [AC_CHECK_LIB(tinfo, tgetent, bash_cv_termcap_lib=libtinfo,
+     [AC_CHECK_LIB(termcap, tgetent, bash_cv_termcap_lib=libtermcap,
         [AC_CHECK_LIB(curses, tgetent, bash_cv_termcap_lib=libcurses,
 	    [AC_CHECK_LIB(ncursesw, tgetent, bash_cv_termcap_lib=libncursesw,
                 [AC_CHECK_LIB(ncurses, tgetent, bash_cv_termcap_lib=libncurses,
--- configure
+++ configure
@@ -7141,14 +7141,14 @@ if test "x$ac_cv_func_tgetent" = xyes
 then :
   bash_cv_termcap_lib=libc
 else case e in #(
-  e) { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for tgetent in -ltermcap" >&5
-printf %s "checking for tgetent in -ltermcap... " >&6; }
-if test ${ac_cv_lib_termcap_tgetent+y}
+  e) { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for tgetent in -ltinfo" >&5
+printf %s "checking for tgetent in -ltinfo... " >&6; }
+if test ${ac_cv_lib_tinfo_tgetent+y}
 then :
   printf %s "(cached) " >&6
 else case e in #(
   e) ac_check_lib_save_LIBS=$LIBS
-LIBS="-ltermcap  $LIBS"
+LIBS="-ltinfo  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -7172,9 +7172,9 @@ return tgetent ();
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
 then :
-  ac_cv_lib_termcap_tgetent=yes
+  ac_cv_lib_tinfo_tgetent=yes
 else case e in #(
-  e) ac_cv_lib_termcap_tgetent=no ;;
+  e) ac_cv_lib_tinfo_tgetent=no ;;
 esac
 fi
 rm -f core conftest.err conftest.$ac_objext conftest.beam \
@@ -7182,20 +7182,20 @@ rm -f core conftest.err conftest.$ac_obj
 LIBS=$ac_check_lib_save_LIBS ;;
 esac
 fi
-{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_termcap_tgetent" >&5
-printf "%s\n" "$ac_cv_lib_termcap_tgetent" >&6; }
-if test "x$ac_cv_lib_termcap_tgetent" = xyes
+{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_tinfo_tgetent" >&5
+printf "%s\n" "$ac_cv_lib_tinfo_tgetent" >&6; }
+if test "x$ac_cv_lib_tinfo_tgetent" = xyes
 then :
-  bash_cv_termcap_lib=libtermcap
+  bash_cv_termcap_lib=libtinfo
 else case e in #(
-  e) { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for tgetent in -ltinfo" >&5
-printf %s "checking for tgetent in -ltinfo... " >&6; }
-if test ${ac_cv_lib_tinfo_tgetent+y}
+  e) { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for tgetent in -ltermcap" >&5
+printf %s "checking for tgetent in -ltermcap... " >&6; }
+if test ${ac_cv_lib_termcap_tgetent+y}
 then :
   printf %s "(cached) " >&6
 else case e in #(
   e) ac_check_lib_save_LIBS=$LIBS
-LIBS="-ltinfo  $LIBS"
+LIBS="-ltermcap  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -7219,9 +7219,9 @@ return tgetent ();
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
 then :
-  ac_cv_lib_tinfo_tgetent=yes
+  ac_cv_lib_termcap_tgetent=yes
 else case e in #(
-  e) ac_cv_lib_tinfo_tgetent=no ;;
+  e) ac_cv_lib_termcap_tgetent=no ;;
 esac
 fi
 rm -f core conftest.err conftest.$ac_objext conftest.beam \
@@ -7229,11 +7229,11 @@ rm -f core conftest.err conftest.$ac_obj
 LIBS=$ac_check_lib_save_LIBS ;;
 esac
 fi
-{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_tinfo_tgetent" >&5
-printf "%s\n" "$ac_cv_lib_tinfo_tgetent" >&6; }
-if test "x$ac_cv_lib_tinfo_tgetent" = xyes
+{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_termcap_tgetent" >&5
+printf "%s\n" "$ac_cv_lib_termcap_tgetent" >&6; }
+if test "x$ac_cv_lib_termcap_tgetent" = xyes
 then :
-  bash_cv_termcap_lib=libtinfo
+  bash_cv_termcap_lib=libtermcap
 else case e in #(
   e) { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for tgetent in -lcurses" >&5
 printf %s "checking for tgetent in -lcurses... " >&6; }


--- examples/rlfe/pty.c
+++ examples/rlfe/pty.c
@@ -248,7 +248,7 @@ OpenPTY(char **ttyn)
 #if defined(HAVE_GETPT) && defined(linux)
   int getpt __P((void));
 #endif
-  sigret_t (*sigcld)__P(SIGPROTOARG);
+  sighandler_t sigcld;
 
   strcpy(PtyName, "/dev/ptmx");
 #if defined(HAVE_GETPT) && defined(linux)


--- terminal.c
+++ terminal.c
@@ -102,15 +102,6 @@ static char *term_string_buffer = (char
 
 static int tcap_initialized;
 
-/* Systems for which PC/BC/UP are defined in the curses library and need an
-   extern definition here. */
-#if !defined (__linux__) && !defined (__gnu_hurd__) && !defined (NCURSES_VERSION)
-#  if defined (__EMX__) || defined (NEED_EXTERN_PC)
-extern
-#  endif /* __EMX__ || NEED_EXTERN_PC */
-char PC, *BC, *UP;
-#endif /* !__linux__ && !NCURSES_VERSION */
-
 
 /* Some strings to control terminal actions.  These are output by tputs (). */
 char *_rl_term_clreol;


--- display.c
+++ display.c
@@ -783,7 +783,7 @@ _rl_optimize_redisplay (void)
 
 /* Useful shorthand used by rl_redisplay, update_line, rl_move_cursor_relative */
 #define INVIS_FIRST()	(local_prompt_invis_chars[0])
-#define WRAP_OFFSET(line, offset)  ((line <= prompt_last_screen_line) ? local_prompt_invis_chars[line] : 0)
+#define WRAP_OFFSET(line, offset)  ((line <= prompt_last_screen_line && local_prompt_invis_chars) ? local_prompt_invis_chars[line] : 0)
 
 #define W_OFFSET(line, offset) ((line) == 0 ? offset : 0)
 #define VIS_LLEN(l)	((l) > _rl_vis_botlin ? 0 : (vis_lbreaks[l+1] - vis_lbreaks[l]))
